include/revng/Yield/Support/GraphLayout/SugiyamaStyle/InternalGraph.h

```diff
@@ -0,0 +1,234 @@
+#pragma once
+
+//
+// This file is distributed under the MIT License. See LICENSE.md for details.
+//
+
+#include <unordered_map>
+#include <variant>
+
+#include "revng/ADT/GenericGraph.h"
+#include "revng/Yield/Support/GraphLayout/Traits.h"
+
+namespace yield::layout::sugiyama {
+
+class InternalGraph;
+
+namespace detail {
+
+struct InternalNode {
+  layout::Point Center;
+  layout::Size Size;
+
+  bool IsVirtual;
+
+private:
+  std::size_t Index;
+
+private:
+  friend InternalGraph;
+  InternalNode(std::size_t Index, layout::Size &&Size) :
+    Center({ 0, 0 }), Size(std::move(Size)), IsVirtual(false), Index(Index) {}
+  InternalNode(std::size_t Index) :
+    Center({ 0, 0 }), Size({ 0, 0 }), IsVirtual(true), Index(Index) {}
+
+public:
+  InternalNode(const InternalNode &) = delete;
+  InternalNode(InternalNode &&) = default;
+  InternalNode &operator=(InternalNode &&) = default;
+
+  std::size_t index() const { return Index; }
+};
+
+struct InternalEdge {
+  bool IsRouted;
+  bool IsBackwards;
+
+private:
+  std::size_t Index;
+
+  using PathOwnership = std::unique_ptr<layout::Path>;
+  using PathReference = layout::Path *;
+  std::variant<PathOwnership, PathReference> Path;
```

Comment:
These could be two separate field, but we never need both of them at the same time - an edge either owns a path, or links to one owned by another edge.

And since converting these two fields into a `variant` brings barely any complexity, there's not much reason not to do it here.

Reply:
Ok but what are the two scenarios where an edge owns the path and where it doesn't?
Can you describe all these things in comments that document this data structure?
Otherwise these subtleties are not self-evident for someone looking at the code for the first time.
Take this as a general rule, when you do a non-obvious design decision, anyone reading your code will be left wondering, so it's your job to comment and document your design decisions.

Reply:
What do you think?
```cpp
/// Represents an edge in the internal graph.
///
/// Because the layouter heavily relies on splitting edges into pieces
/// (partitioning them, see `GraphPreparation.cpp` for more details), we need
/// a way to let multiple paths accumulate into a single path that can
/// eventually be exported for the original edge.
///
/// To achieve this, each edge holds EITHER a real path (`PathOwner`) OR just
/// a reference to a path held by some other edge (`PathReference`).
///
/// To limit the opportunities for breaking stuff by using this non-intuitive
/// system incorrectly, access to the path is heavily limited to just two
/// methods:
/// - `appendPoint` - adds a point to the current path without regard of whether
///                   the path is owned by this edge, or some other one.
/// - `getPath`     - returns the path reference (which allows it to be moved
///                   out), but is only defined for the edge owners. This
///                   guarantees that each path can only be returned at most
///                   once (the only accessor asserts it).
///
/// There's also a `isVirtual` helper for determining whether the path is owned
/// or not. The expected usage pattern is
///
/// for (auto &Edge : allTheEdgesOfMyGraph())
///   if (Edge.isVirtual())
///     acceptThePath(std::move(Edge.getPath()));
```
(this doxygen is attached to `InternalEdge` class definition).

Reply:
I'm closing this as per the mattermost discussion.